src/NuGet.Core/NuGet.Protocol/NameSpace/SearchTree.cs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.Protocol
+{
+    public class SearchTree : INameSpaceLookup
+    {
+        private readonly SearchNode _root;
+        private const int PackageIdMaxLength = 100;
+
+        public SearchTree(IReadOnlyList<PackageSourceSection> packageSourceSections)
+        {
+            if (packageSourceSections == null)
+            {
+                throw new ArgumentNullException(nameof(packageSourceSections));
+            }
+
+            _root = new SearchNode(null);
+
+            foreach (PackageSourceSection packageSourceSection in packageSourceSections)
+            {
+#pragma warning disable CA1308 // Normalize strings to uppercase
```

Comment:
I kinda hate this warning, CA1308 . 

On our end, it actually creates more false positives than true positives.